Usbipd/wsl-scripts/auto-attach.sh

```diff
@@ -96,7 +96,7 @@ is_attached() {
 sleep() {
     local SECONDS=$(($1))
     # sleep without creating a new process
-    read -r -t $SECONDS </proc/self/fd/1 || :
+    read -r -t $SECONDS || :
```

Comment:
```suggestion
    read -r -t $SECONDS || /usr/bin/sleep $SECONDS
```

Some more info. It turns out that the original code (reading from stdout) is actually the same as reading from stdin. I.e., the "trick" that I tried to use (guaranteed no input available, yet not EOF) isn't working anyway. So, your fix is good in principle. However, if stdin is closed (which may be the case for systemd), then read will return immediately (with a non-0 exit). This would lead to a busy loop. My suggestion is to fall back to regular sleep in that case. At the cost of creating a new process for every sleep. At least that is better than a busy loop.

Reply:
Seems you are right, this indeed creates a busy loop, so perhaps we could check if stdin/stdout is available and then use this trick, otherwise just fallback to sleep

Reply:
I found another quite sneaky hack and that was to use a FIFO pipe. E.g. put something like this at the top of the script so it only runs once:
```sh
sleeppipe="$(mktemp -u)"
mkfifo -m 600 "$sleeppipe"
exec 3<>"$sleeppipe"
trap 'rm "$sleeppipe"' EXIT
```

Then in the sleep command:
```sh
read -r -t $SECONDS <"$sleeppipe" || :
```

I tested that this worked both inside and outside of systemd. But perhaps it is too much trickery to be worth it?

Reply:
Yes, I know about that one. Unfortunately it is a named (filesystem) pipe. What we really want is an anonymous pipe. But bash has none, except via `coproc`. However, that creates a sub-process (only one though, at startup), whose lifetime is not manageable.

All in all, I would just fallback to regular `sleep`. One thing to remember is that this needs to be distro-independent.

One further improvement would be to call our "optimized sleep" function different, so we can just fallback to `sleep` instead of `/usr/bin/sleep`. That would work on weird distros that have sleep somewhere else in PATH.

Who will do the extra commit: you or me?

PS: thanks for contributing and making improvements!

Reply:
Thank you as well for the reflection and investigation, I'm on one of those "weird distros", so I would suggest relying on `sleep` from `PATH` instead of fixed path. I'm fine with you doing it, since it is not my original change, I don't think it provides much value me doing it. Unless it is easier for you of course, then I could do the change :)

Reply:
There is a problem with:
```
read -r -t $SECONDS || /usr/bin/sleep $SECONDS
```

Seems that the timeout of read will also trigger a falsy status and then enter the else condition...

Reply:
I took another approach, I checked for fd 0 and only defines the custom sleep function if it is available, otherwise it will fall-back to the sleep command.

Reply:
How about this one?

```bash
sleep() {
  local SECONDS=$(($1))
  local ERROR=0
  read -t $SECONDS || ERROR=$?
  if [[ $ERROR -le 128 && $ERROR -ne 0 ]]; then
    # neither timeout nor success, use regular sleep instead
    command sleep $SECONDS
  fi
}
```

This allows the user to hit `<enter>` to try immediately (read will return 0). Detects proper (intended) timeout. And anything else (including non-existing or closed file descriptors) falls back to normal sleep command.

Reply:
I can confirm that this solution works as well, so you choose which one you want to go with. Either your proposed one, or only defining the sleep function when it can use the custom read solution.

Reply:
I went with my latest version, but merged yours first. So, now you are on the contributors list.

Thanks for contributing!